src/main/java/com/algorand/algosdk/util/abi/values/Value.java

```diff
@@ -0,0 +1,171 @@
+package com.algorand.algosdk.util.abi.values;
+
+import com.algorand.algosdk.util.Encoder;
+import com.algorand.algosdk.util.abi.types.*;
+
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class Value {
+    public Type abiType;
+    public Object value;
+
+    Value() {
+    }
+
+    private Value(Value o) {
+        this.abiType = o.abiType;
+        this.value = o.value;
+    }
+
+    public abstract byte[] encode();
+
+    public Value(byte[] encoded, Type abiType) {
+        this(Value.decode(encoded, abiType));
+    }
+
+    public static Value decode(byte[] encoded, Type abiType) {
```

Comment:
Can `decode` be declared as an `abstract` method here and be overriden in the child classes? It could help eliminate any switching or abiType checks, and we could probably combine the `TupleDecode` functions.

Reply:
I had another idea. We can probably construct from the bytes and essential components for type.
In this case, the switching can be eliminated, and type checking is done while constructing type elements, the decoding is then proceed with new constructed type elements...

Seems to be fitting your thoughts?

Reply:
That's true -- I like the idea!